staking-contract/contracts/StakingRewardSystem.sol

```diff
@@ -0,0 +1,221 @@
+    event RewardClaimed(address indexed user, uint256 rewardAmount, address indexed rewardToken);
+    event PenaltiesWithdrawn(address indexed owner, uint256 amount);
+    event RewardTokenAdded(address indexed rewardToken, address addedBy);
+    event MinStakingPeriodUpdated(uint256 previousPeriod, uint256 newPeriod, address updatedBy);
+    event EarlyWithdrawalPenaltyUpdated(uint256 previousPenalty, uint256 newPenalty, address updatedBy);
+    event RewardRateUpdated(uint256 previousRate, uint256 newRate, address updatedBy);
+
+
+    // Custom errors for more efficient error handling
+    error InvalidAmount();
+    error InvalidRewardToken();
+    error WithdrawAmountExceedsStake();
+    error NoStakeFound();
+    error CannotClaimRewardYet();
+    error NoPenaltiesToWithdraw();
+    error InvalidInput(string message); 
+    error UserNotExist();  
+
+    // Initializer function to replace constructor for upgradeable contracts
+    function initialize(address _stakingToken, address[] memory _rewardTokens) public initializer {
+        if (_rewardTokens.length == 0) revert InvalidInput("At least one reward token must be provided");
+        if (_stakingToken == address(0)) revert InvalidInput("Staking token address cannot be zero");
+        __Ownable_init();
+        __Pausable_init();
+        stakingToken = IERC20(_stakingToken);
+
+        minStakingPeriod = 30 *86400;
+        earlyWithdrawalPenalty = 1e16;
+        rewardRate = 1 * 1e18;
+
+        for (uint256 i = 0; i < _rewardTokens.length; i++) {
+            _addRewardToken(_rewardTokens[i]);
+        }
+    }
+
+    // Pause the contract
+    function pause() external onlyOwner {
+        _pause();
+    }
+
+    // Unpause the contract
+    function unpause() external onlyOwner {
+        _unpause();
+    }
+
+    // Function to stake tokens
+    function stake(uint256 _amount, address _rewardToken) external nonReentrant whenNotPaused {
+        if (_amount <= 0) revert InvalidAmount();
+        if (_rewardToken == address(0)) revert InvalidInput("Reward token address cannot be zero");
+        if (!isRewardToken[_rewardToken]) revert InvalidRewardToken();
+        stakingToken.safeTransferFrom(msg.sender, address(this), _amount);
+        Stake storage userStake = userStakes[msg.sender][_rewardToken];
+        userStake.amount += _amount;
+        userStake.startTime = block.timestamp;
+        emit Staked(msg.sender, _amount, _rewardToken);
+    }
+
+    // Function to withdraw staked tokens
+    function withdraw(uint256 _amount, address _rewardToken) external nonReentrant whenNotPaused {
+        if (_amount <= 0) revert InvalidAmount();
+        if (_rewardToken == address(0)) revert InvalidInput("Reward token address cannot be zero");
+        if (!isRewardToken[_rewardToken]) revert InvalidRewardToken();
+        Stake storage userStake = userStakes[msg.sender][_rewardToken];
```

Comment:
This code does not follow DRY principle as you already using the same code in stake function. Better create a seperate common function for initial valid amount checks and try to reuse the code as much as possible. 
Ask ChatGPT to remove redundant code and reuse the code.